### DISCUSSION TITLE
Como fazer o app conseguir pegar os dados do banco de dados pela API já feita?

### DISCUSSION THREAD

**Discussion by vinivp1** [Flutterando/forum#196] [Perguntas e Respostas]
na pasta user_repository o meu código está assim, pegando o resultado que a API me dá e transformando do json para o normal:

`class UserRepository extends GetConnect{
  Future<List<User>> getUserList()async{

    List<User> userList = [];

    var path = 'https://flasktest-f17y.onrender.com/login?argumentos=';

    final response = await get(path);

    final users = response.body;

    for(Map user in users){
      User u = User.fromJson(user);

      userList.add(u);
    }

    return userList;
  }
}

class PasswordRepository extends GetConnect{
  Future<List<Password>> getPasswordList()async{

    List<Password> passwordList = [];

    var path = 'https://flasktest-f17y.onrender.com/login?argumentos=';

    final response = await get(path);

    final passwords = response.body;

    for(Map password in passwords){
      Password u = Password.fromJson(password);

      passwordList.add(u);
    }

    return passwordList;
  }
}`

Creio eu que o erro esteja nas linhas de códigos a seguir, onde era pra se checar email por email vindo da API e se tiver o email no banco de dados comparar com a senha cadastrada:

`class LoginController extends GetxController {
  UserRepository repository = UserRepository();
  PasswordRepository rep = PasswordRepository();

  Future<List<User>> userList() async => await repository .getUserList();
  Future<List<Password>> passwordList() async => await rep .getPasswordList();

  TextEditingController emailInput = TextEditingController();
  TextEditingController passwordInput = TextEditingController();

  User newUser = User();

  @override
  void onInit() {
    super.onInit();

    userList().then((users) {
      passwordList().then((passwords) {
        for (int i = 0; i < users.length; i++) {
          if (emailInput.text == users[i].email) {
            checkPassword(i, passwords);
            break;
          }
        }
      }).catchError((error) {
        printError(error.toString());
      });
    }).catchError((error) {
      printError(error.toString());
    });
  }

  void checkPassword(int i, List<Password> passwords) {
    if(passwordInput.text == passwords[i].password){
      login();
    } else {
      printError("Senha Incorreta");
    }
  }

  void login() {
    Get.to(HomeView());
  }

  void printError(String error) {
    print(error);
  }
}`

Já pesquisei em vários lugares mas não consigo tirar minha dúvida, espero que alguém consiga me ajudar, obrigado desde já!!
OBS: Não programo em flutter este é só um projeto para complementar um trabalho da faculdade, estávamos programando pelo kodular.io onde a programação é em bloco mas decidimos mudar para o flutter.

**Comment by rudsonalves** [comment]
O problema no seu código está em várias áreas que podem causar falhas na lógica e na execução do processo de login. Vou detalhar os principais problemas e como eles foram resolvidos:

### 1. Tratamento de Resposta da API:

O código não está verificando se a resposta da API está retornando corretamente ou se há erros na requisição. Além disso, não está garantindo que o corpo da resposta (`response.body`) seja tratado corretamente como uma lista de mapas.

Adicione verificações de erro (`response.status.hasError`) e assegure que a resposta seja tratada como uma lista de mapas (`final users = response.body as List`).

```dart
if (response.status.hasError) {
  print('Error: ${response.statusText}');
  return Future.error(response.statusText ?? 'Unknown error');
}

final users = response.body as List;
```

### 2. Lógica de Verificação do Usuário e Senha:

A lógica dentro de `onInit` para iterar sobre as listas de usuários e senhas não está clara e podia causar problemas de desempenho e lógica.

Mude-a para usar métodos como `firstWhereOrNull` do pacote `collection`, que torna a busca por um usuário e sua senha mais clara e eficiente.

```dart
User? user = users.firstWhereOrNull((u) => u.email == emailInput.text);
if (user != null) {
  Password? password = passwords.firstWhereOrNull((p) => p.userId == user.id);
  if (password != null && password.password == passwordInput.text) {
    Get.to(HomeView());
  } else {
    printError("Senha Incorreta");
  }
} else {
  printError("Email não encontrado");
}
```

### 3. Uso Correto de `onInit` e `async/await`:

A lógica dentro de `onInit` está misturando chamadas assíncronas sem a correta estrutura `async/await`, o que pode causar comportamento inesperado.

#### Solução:

Refatore a função `login` para ser chamada explicitamente e use `async/await` corretamente para esperar pelas listas de usuários e senhas.

```dart
void login() async {
  try {
    List<User> users = await userList();
    List<Password> passwords = await passwordList();

    User? user = users.firstWhereOrNull((u) => u.email == emailInput.text);
    if (user != null) {
      Password? password = passwords.firstWhereOrNull((p) => p.userId == user.id);
      if (password != null && password.password == passwordInput.text) {
        Get.to(HomeView());
      } else {
        printError("Senha Incorreta");
      }
    } else {
      printError("Email não encontrado");
    }
  } catch (e) {
    printError(e.toString());
  }
}
```

### 4. Identificação de Usuário:

Não está claro como os dados de senha estão associados aos usuários, pois não há referência de um identificador de usuário (`userId`).

Assegure que a associação entre usuários e senhas seja clara utilizando um identificador de usuário (`userId`) nos dados de senha.

### Resumo:

Esses ajustes garantem que:

1. A resposta da API seja tratada corretamente e os erros sejam capturados e manipulados.
2. A lógica de verificação de usuários e senhas seja eficiente e clara.
3. As chamadas assíncronas sejam usadas corretamente para evitar comportamento inesperado.
4. A associação entre usuários e senhas seja clara e correta.

Esses são os principais problemas que identifiquei no seu código original, e as soluções fornecidas abordam essas questões para tornar o código mais robusto e funcional.
